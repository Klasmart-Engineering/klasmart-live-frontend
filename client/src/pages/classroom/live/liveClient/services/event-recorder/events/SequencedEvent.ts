import { SqeuencedEventBuilder } from '../builder/SequencedEventBuilder'

export interface SequencedEventData {
    streamId: string;
    sequence: number;
    tag: string;
    isKeyFrame: boolean;
    data: string;
}

export class SequencedEvent {
    readonly streamId: string;
    readonly sequence: number;
    readonly tag: string;
    readonly isKeyFrame: boolean;
    readonly data: string;

    private constructor (streamId: string, sequence: number, tag: string, isKeyFrame: boolean, data: string) {
      this.streamId = streamId
      this.sequence = sequence
      this.tag = tag
      this.isKeyFrame = isKeyFrame
      this.data = data
    }

    toData (): SequencedEventData {
      return {
        streamId: this.streamId,
        sequence: this.sequence,
        tag: this.tag,
        isKeyFrame: this.isKeyFrame,
        data: this.data
      }
    }

    static fromData (data: SequencedEventData) : SequencedEvent {
      return new SequencedEvent(data.streamId, data.sequence, data.tag, data.isKeyFrame, data.data)
    }

    static builder (sequencedEvent?: SequencedEvent) : SqeuencedEventBuilder {
      return new SqeuencedEventBuilder(sequencedEvent)
    }
}
