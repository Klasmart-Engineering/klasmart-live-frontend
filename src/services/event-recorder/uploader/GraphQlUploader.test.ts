import {
    GraphQlUploader,
    mapEventData,
} from './GraphQlUploader';
import { SequencedEventData } from '@/services/event-recorder/events/SequencedEvent';
import {
    ClientError,
    GraphQLClient,
} from 'graphql-request';
import { mocked } from 'jest-mock';

const data: SequencedEventData[] = [
    {
        streamId: `streamId`,
        sequence: 1,
        isKeyFrame: true,
        tag: `tag`,
        data: JSON.stringify({
            checkout: true,
            event: `checkout`,
            index: 0,
        }),
    },
];
const url = `http://localhost:8000/graphql`;
const mutation = `mutation`;
const refreshEndpoint = `refreshEndpoint`;

// The following error comes from the GraphQL server.
// If server is updated, this will need to be updated.
const authenticationError = Object.create(ClientError.prototype);
authenticationError.response = {
    errors: [
        {
            message: `Context creation failed: Error: Missing JWT token`,
            extensions: {
                code: `AuthenticationInvalid`,
            },
        },
    ],
    status: 400,
};

let postedErrorMessage: unknown;
Object.defineProperty(global.window.parent, `postMessage`, {
    value: (message: unknown) => {
        postedErrorMessage = message;
    },
});

let client = new GraphQLClient(url);
let mockRequest = jest.spyOn(client, `request`);
let uploader = new GraphQlUploader(client, mutation, refreshEndpoint);

describe(`GraphQlUploader`, () => {
    beforeEach(() => {
        jest.clearAllMocks();
        client = new GraphQLClient(url);
        mockRequest = jest.spyOn(client, `request`);
        uploader = new GraphQlUploader(client, mutation, refreshEndpoint);
        postedErrorMessage = undefined;
    });

    afterAll(() => {
        jest.restoreAllMocks();
    });

    test(`should upload event data`, async () => {
        const mappedData = data.map(mapEventData);
        mockRequest.mockResolvedValue(undefined);
        const uploadPromise = uploader.upload(data);
        await expect(uploadPromise).resolves.toBeUndefined();
        expect(mockRequest)
            .toHaveBeenCalledTimes(1);
        expect(mockRequest)
            .toHaveBeenCalledWith(mutation, {
                streamId: data[0].streamId,
                pageEvents: mappedData,
            });
    });

    test(`should handle error`, async () => {
        const mockError = mocked(ClientError, true);
        mockRequest.mockRejectedValue(mockError);
        const uploadPromise = uploader.upload(data);
        await expect(uploadPromise).rejects.toBe(mockError);
    });

    test(`should handle authentication error by making a call to refresh endpoint`, async () => {
        mockRequest.mockRejectedValue(authenticationError);
        global.fetch = jest.fn(() =>
            Promise.resolve({
                ok: true,
            })) as jest.Mock;

        const uploadPromise = uploader.upload(data);
        await expect(uploadPromise).rejects.toBe(authenticationError);
        expect(global.fetch)
            .toHaveBeenCalledTimes(1);
        expect(global.fetch)
            .toHaveBeenCalledWith(refreshEndpoint, {
                credentials: `include`,
            });
        expect(postedErrorMessage)
            .toBeUndefined();
    });

    test(`should handle unsuccessful refresh attempt during authentication error`, async () => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                ok: false,
            })) as jest.Mock;

        mockRequest.mockRejectedValue(authenticationError);
        const uploadPromise = uploader.upload(data);
        await expect(uploadPromise).rejects.toBe(authenticationError);
        expect(postedErrorMessage)
            .toBeDefined();
        expect(postedErrorMessage)
            .toMatchSnapshot();
    });
});
