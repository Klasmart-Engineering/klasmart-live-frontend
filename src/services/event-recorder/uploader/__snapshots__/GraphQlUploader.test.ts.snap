// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQlUploader should handle unsuccessful refresh attempt during authentication error 1`] = `
Object {
  "error": "RedirectToLogin",
}
`;
